Chore/browser tests

ryan-roemer opened the pull request:
**Status**: Ready for review!

- Refactor the test directory structure.
- Imports the `node` tests into the browser, building with webpack. They are the source of truth.
- Run headless chrome in Travis.
- Run IE9 in Appveyor.
- Use `builder` to parallelize running lint and various tests.
- Fixes #7 

Stuff left to do:

- Code coverage: https://github.com/FormidableLabs/react-fast-compare/issues/11
- Sauce labs for more coverage: https://github.com/FormidableLabs/react-fast-compare/issues/12

/cc @chrisbolin

ryan-roemer commented on .travis.yml:
```diff
@@ -1,8 +1,37 @@
 language: node_js
+
 node_js:
-  - "4"
```
node4 is EOL'ed in basically two weeks: https://github.com/nodejs/Release#release-schedule

ryan-roemer commented on .travis.yml:
```diff
@@ -1,8 +1,37 @@
   - coveralls < coverage/lcov.info
+
+cache:
+  yarn: true
```
I believe so -- just as you have a `$HOME` based yarn cache specific to yarn.

ryan-roemer replied:
Update -- this is the preferred caching config per Travis: https://blog.travis-ci.com/2016-11-21-travis-ci-now-supports-yarn#caching-with-yarn

ryan-roemer commented on package.json:
```diff
@@ -5,12 +5,14 @@
-    "test-spec-watch": "mocha spec/*.spec.js -R spec --watch",
-    "test-spec": "mocha spec/*.spec.js -R spec",
-    "test-cov": "nyc npm run test-spec",
+    "eslint": "eslint \"*.js\" benchmark test",
```
Everything is now windows compliant. Not sure if this specific one was necessary, but generally needed.

chrisbolin replied:
ah! amazing how easy it is to forget that one non-*nix system 😂

ryan-roemer commented on package.json:
```diff
@@ -5,12 +5,14 @@
   "scripts": {
     "benchmark": "node benchmark",
-    "eslint": "eslint *.js benchmark spec",
-    "test-spec-watch": "mocha spec/*.spec.js -R spec --watch",
```
I'd prefer we just document:

```hs
$ yarn test-node --watch
```

Rather than multiplying the number of permutations in `package.json:scripts`

Side note -- I believe the best normal way to view developing is "just run `yarn test-node`" and saving the browser tests for before a PR / when you're doing something that might drastically change from browser-to-browser. But since we're just doing mostly vanillajs, the normal happy path workflow should be fast.

chrisbolin replied:
wouldn't it be `yarn test-node -- --watch`?

ryan-roemer replied:
It would be `npm run test-node -- --watch` or `builder run test-node -- --watch`.

However `yarn` optimizes away the need for both `run` and the "extra flags flag (`--`)" in most cases. In fact, if you _do_ use `--` then you get this helpful message:

```
warning From Yarn 1.0 onwards, scripts don't require "--" for options to be forwarded. In a future version, any explicit "--" will be forwarded as-is to the scripts.
```

chrisbolin commented:
looks good. so let's just 
- fix `yarn benchmark`
- document `yarn test-node -- --watch`

chrisbolin commented on package.json:
```diff
@@ -5,12 +5,14 @@
   "main": "index.js",
   "scripts": {
     "benchmark": "node benchmark",
```
we'll have to change the directory that `benchmark/index.js` gets `tests.js` now that it moved

ryan-roemer commented:
@chrisbolin -- I think I've addressed everything in: https://github.com/FormidableLabs/react-fast-compare/pull/10/commits/ee3946abd552934345b9bcc5b8e4024b7727c9ae

Also note that I've added `npm version` as our designated workflow because all our other projects do it and it's a good way to ensure the tests are run before publishing ;)

chrisbolin commented:
🎉
